// ampersand-sync/ampersand-sync.js
/*$AMPERSAND_VERSION*/
var _ = require('underscore');
var xhr = require('xhr');
var qs = require('qs');


// Throw an error when a URL is needed, and none is supplied.
var urlError = function() {
    throw new Error('A "url" property or function must be specified');
};


module.exports = function(method, model, options) {
    var type = methodMap[method];
    var headers = {};
    var records = (localStorage.getItem(name) && localStorage.getItem(name).split(',')) || [];

    // Default options, unless specified.
    _.defaults(options || (options = {}), {
        emulateHTTP: false,
        emulateJSON: false,
        // overrideable primarily to enable testing
        xhrImplementation: xhr
    });

    // Default request options.
    var params = {
        type: type
    };

    // Ensure that we have a URL.
    if (!options.url) {
        options.url = _.result(model, 'url') || urlError();
    }

    // Ensure that we have the appropriate request data.
    if (options.data == null && model && (method === 'create' || method === 'update' || method === 'patch')) {
        params.json = options.attrs || model.toJSON(options);
    }

    // If passed a data param, we add it to the URL or body depending on request type
    if (options.data && type === 'GET') {
        // make sure we've got a '?'
        options.url += _.contains(options.url, '?') ? '&' : '?';
        options.url += qs.stringify(options.data);
    }

    // Set the data from the local storage in the model
    if (localStorage.getItem(options.url) && type!=='POST')
        options.success(JSON.parse(localStorage.getItem(options.url)), 'success');

    // For older servers, emulate JSON by encoding the request into an HTML-form.
    if (options.emulateJSON) {
        headers['Conent-Type'] = 'application/x-www-form-urlencoded';
        params.body = params.json ? {
            model: params.json
        } : {};
        delete params.json;
    }

    // For older servers, emulate HTTP by mimicking the HTTP method with `_method`
    // And an `X-HTTP-Method-Override` header.
    if (options.emulateHTTP && (type === 'PUT' || type === 'DELETE' || type === 'PATCH')) {
        params.type = 'POST';
        if (options.emulateJSON) params.body._method = type;
        headers['X-HTTP-Method-Override'] = type;
    }

    // When emulating JSON, we turn the body into a querystring.
    // We do this later to let the emulateHTTP run its course.
    if (options.emulateJSON) {
        params.body = qs.stringify(params.body);
    }

    // Start setting ajaxConfig options (headers, xhrFields).
    var ajaxConfig = (_.result(model, 'ajaxConfig') || {});

    // Combine generated headers with user's headers.
    if (ajaxConfig.headers) {
        _.extend(headers, ajaxConfig.headers);
    }
    params.headers = headers;

    //Set XDR for cross domain in IE8/9
    if (ajaxConfig.useXDR) {
        params.useXDR = true;
    }

    // Set raw xhr options.
    if (ajaxConfig.xhrFields) {
        var beforeSend = ajaxConfig.beforeSend;
        params.beforeSend = function(req) {
            for (var key in ajaxConfig.xhrFields) {
                req[key] = ajaxConfig.xhrFields[key];
            }
            if (beforeSend) return beforeSend.apply(this, arguments);
        };
        params.xhrFields = ajaxConfig.xhrFields;
    } else {
        params.beforeSend = ajaxConfig.beforeSend;
    }

    // Turn a jQuery.ajax formatted request into xhr compatible
    params.method = params.type;

    var ajaxSettings = _.extend(params, options);

    var self = this;

    // Make the request. The callback executes functions that are compatible
    // With jQuery.ajax's syntax.
    var request = options.xhr = options.xhrImplementation(ajaxSettings, function(err, resp, body) {
        if (err && options.error) return options.error(resp, 'error', err.message);

        // Parse body as JSON if a string.
        if (body && typeof body === 'string') {
            try {
                body = JSON.parse(body);

                switch (method) {
                    case 'create':
                        if (body && !body.id) body.id = guid();
                        localStorage.setItem(options.url, resp.rawRequest.response);
                        break;
                    case 'update':
                        var body = JSON.parse(resp.rawRequest.response);
                        if (body && !body.id) body.id = model.id;
                        localStorage.setItem(options.url, resp.rawRequest.response);
                        break;
                    case 'delete':
                        records.splice(records.indexOf(model.id.toString()), 1);
                        localStorage.removeItem(options.url);
                        break;
                    case 'read':
                        localStorage.setItem(options.url, resp.body);
                        break;
                }
            } catch (err) {
                if (options.error) return options.error(resp, 'error', err.message);
            }
        }
        if (options.success) return options.success(body, 'success', resp);
    });

    model.trigger('request', model, request, options, ajaxSettings);
    request.ajaxSettings = ajaxSettings;

    return request;
};

function guid() {
    function S4() {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    }
    return S4() + S4();
}

// Map from CRUD to HTTP for our default `Backbone.sync` implementation.
var methodMap = {
    'create': 'POST',
    'update': 'PUT',
    'patch': 'PATCH',
    'delete': 'DELETE',
    'read': 'GET'
};
